capplets/appearance/appearance-font.c

```diff
@@ -131,203 +110,80 @@ typedef enum {
-	ascent = 0;
-	descent = 0;
-	
-	if (font1)
-	{
-		XftTextExtentsUtf8 (xdisplay, font1, (unsigned char*) string1,
-		strlen (string1), &extents1);
-		ascent = MAX (ascent, font1->ascent);
-		descent = MAX (descent, font1->descent);
-	}
 
-	if (font2)
-	{
-		XftTextExtentsUtf8 (xdisplay, font2, (unsigned char*) string2, strlen (string2), &extents2);
-		ascent = MAX (ascent, font2->ascent);
-		descent = MAX (descent, font2->descent);
-	}
+	context = gtk_widget_get_pango_context (darea);
+	set_fontoptions (context, antialiasing, hinting);
+	layout = pango_layout_new (context);
 
-	width = extents1.xOff + extents2.xOff + 4;
-	height = ascent + descent + 2;
+	fd = pango_font_description_from_string ("Serif");
+	pango_layout_set_font_description (layout, fd);
+	pango_font_description_free (fd);
 
-#if !GTK_CHECK_VERSION (3, 0, 0)
-	pixmap = gdk_pixmap_new (NULL, width, height, gdk_visual_get_depth (visual));
-#endif
+	pango_layout_set_markup (layout, str, -1);
 
-#if GTK_CHECK_VERSION (3, 0, 0)
-	draw = XftDrawCreate (xdisplay, GDK_WINDOW_XID (gdk_screen_get_root_window (gdk_screen_get_default ())), xvisual, xcolormap);
-#else
-	draw = XftDrawCreate (xdisplay, GDK_DRAWABLE_XID (pixmap), xvisual, xcolormap);
-#endif
+	pango_layout_get_extents (layout, NULL, &extents);
+	width = PANGO_PIXELS(extents.width) + 4;
+	height = PANGO_PIXELS(extents.height) + 2;
 
-	rendcolor.red = 0;
-	rendcolor.green = 0;
-	rendcolor.blue = 0;
-	rendcolor.alpha = 0xffff;
-	
-	XftColorAllocValue(xdisplay, xvisual, xcolormap, &rendcolor, &black);
+	surface = cairo_image_surface_create (CAIRO_FORMAT_A8, width, height);
+	cr = cairo_create (surface);
 
-	rendcolor.red = 0xffff;
-	rendcolor.green = 0xffff;
-	rendcolor.blue = 0xffff;
-	rendcolor.alpha = 0xffff;
-	
-	XftColorAllocValue(xdisplay, xvisual, xcolormap, &rendcolor, &white);
-	XftDrawRect(draw, &white, 0, 0, width, height);
-	
-	if (font1)
-	{
-		XftDrawStringUtf8(draw, &black, font1, 2, 2 + ascent, (unsigned char*) string1, strlen(string1));
-	}
-	
-	if (font2)
-	{
-		XftDrawStringUtf8(draw, &black, font2, 2 + extents1.xOff, 2 + ascent, (unsigned char*) string2, strlen(string2));
-	}
-
-	XftDrawDestroy(draw);
-
-	if (font1)
-	{
-		XftFontClose(xdisplay, font1);
-	}
-	
-	if (font2)
-	{
-		XftFontClose(xdisplay, font2);
-	}
-
-#if GTK_CHECK_VERSION (3, 0, 0)
-	tmp_pixbuf = gdk_pixbuf_new (GDK_COLORSPACE_RGB, TRUE,8, width, height);
-#else
-	tmp_pixbuf = gdk_pixbuf_get_from_drawable(NULL, pixmap, colormap, 0, 0, 0, 0, width, height);
-#endif
-	pixbuf = gdk_pixbuf_scale_simple(tmp_pixbuf, 1 * width, 1 * height, GDK_INTERP_TILES);
-
-#if !GTK_CHECK_VERSION (3, 0, 0)
-	g_object_unref(pixmap);
-#endif
-	g_object_unref(tmp_pixbuf);
+	pango_cairo_show_layout (cr, layout);
+	g_object_unref (layout);
+	cairo_destroy (cr);
 
-	g_object_set_data_full(G_OBJECT(darea), "sample-pixbuf", pixbuf, (GDestroyNotify) g_object_unref);
+	g_object_set_data_full(G_OBJECT(darea), "sample-surface", surface, (GDestroyNotify) cairo_surface_destroy);
```

Comment:
Here the surface and its destroy function are associated to the object, so that when the object is destroyed the surface is also destroyed.
A quick valgrind check shows there is no memory leak around here.